test/blackbox/common/DDSBlackboxTestsContentFilter.cpp

```diff
@@ -610,6 +610,188 @@ TEST(DDSContentFilter, CorrectlyHandleAliasOtherHeader)
     EXPECT_NE(nullptr, filtered_topic);
 }
 
+/*
+ * Regression test for https://eprosima.easyredmine.com/issues/20815
+ * Check that the content filter is only applied to alive changes.
+ * The test creates a reliable writer and a reader with a content filter that only accepts messages with a specific
+ * string. After discovery, the writer sends 10 samples which pass the filer in 10 different instances, with the
+ * particularity that after each write, the instance is unregistered.
+ * The DATA(u) generated would not pass the filter if it was applied. To check that the filter is only applied to
+ * ALIVE changes (not unregister or disposed), the test checks that the reader receives 10 valid samples (one per
+ * sample sent) and 10 invalid samples (one per unregister). Furthermore, it also checks that no samples are lost.writer
+ */
+TEST(DDSContentFilter, OnlyFilterAliveChanges)
```

Comment:
I think it would be worth to try extending the already existing `PubSubReader` to accept `content_filter` stuff (also for future usage) and then inheriting from it to implementing test's logic  i.e `receive_one()` method could be overridden and the samples logic can be placed there. The `PubSubWriter` can be also used in the test. (e.g discovery, writer `matched_`). What do you think ? IMO it would give more simplicity instead of creating everything custom to this specific test (which I do not argue that it may be necessary if extending the `PubSubReader` and `PubSubWriter` gets too tangled). 

NIT, would it be more correct using `take_next_instance` given that the type is keyed ? Although I understand that the use-case scenario may be using `take_next_sample` and this test tries to mimic the same behavior.